src/runtime/runtime.go

```diff
@@ -15,14 +15,27 @@ func initAll()
 //
 // Without scheduler:
 //
-//     main.main()
+//     mainFunc()
 //
 // With scheduler:
 //
-//     main.main()
+//     go wrapMain()
 //     scheduler()
 func callMain()
 
+// mainFunc is a temporary value that will be later replaced with the actual user-provided main function
+func mainFunc()
+
+// wrapMain is a wrapper which is used for invoking main.
+// When main completes, this allows the program to return.
+func wrapMain() {
+	// run main
+	mainFunc()
+
+	// when main is done, let the scheduler exit
+	schedulerDone = true
```

Comment:
It seems like this `schedulerDone` system is not really related to the rest of the PR, is that correct? In that case, it would make reviewing easier if it was split in two: one commit (or PR) for only the `schedulerDone` mechanism and the other a pure refactoring change (no behavioral change).

Reply:
I originally put `schedulerDone` in a seperate PR, but later chose to do it in one PR instead because it would have needed to be merged first, and would have resulted in contradictory changes.

Reply:
The problem is that the `schedulerDone` mechanism (at least as far as I understand it) seems fairly straightforward while the scheduler refactorings seem far more complicated. In particular, I'm missing the bigger context in the refactor: how this helps in implementing wakeups from interrupts (often, an unvalidated design lacks in some hard-to-predict ways and I'd like to avoid that situation).

Splitting would greatly help by getting the first changes in and being able to focus on the other changes. If they conflict, what I usually do is make two PRs: one with the first commit and one with both commit. And when the first commit is merged, I rebase on top of the dev branch to make the diff clean and the PR mergeable (GitHub fails to see that the PR can in fact be merged even though a simple rebase fixes any conflicts).

Reply:
I reopened #633 and rebased on it as requested.

>In particular, I'm missing the bigger context in the refactor: how this helps in implementing wakeups from interrupts (often, an unvalidated design lacks in some hard-to-predict ways and I'd like to avoid that situation).
This is effectively a cleaned-up version of the interrupt queue mechanism from #606. This is the mechanism previously discussed. Essentially, `poll` checks for events, and `wait` runs a `wfi`/`wfe`.